### ISSUE TITLE
restore creates the parent folder with wrong permissions and date

### ISSUE THREAD

**Issue by tessus** [kopia/kopia#4324]
I backup my macOS user folder `/Users/myuser` with kopia every day.

When I restore (via the KopiaUI) the files from `/Users/myuser/testfolder` to `/Users/myuser/testfolder-restore` (I have to specify the target dir in the UI), the `testfolder-restore` folder has the wrong permissions and date. The files in that folder all have the correct permissions and date.

Instead of `0755` it has `0555` and the date is `1969-12-31 19:00`.

Is there a way (UI or command line) to restore `/Users/myuser/testfolder` with correct permissions and timestamp (this information is after all stored correctly in the snapshot)?

**Comment by ndas108** [comment]
Yes, I experience exactly the same issue, which seems to be related to this one: #4066

**Comment by ndas108** [comment]
Restore destination path in the gui (at least in Fedora 41) is also buggy: ~/test/ creates '~/test/' instead of just '/test/', but that's probably a separate issue.

**Comment by tessus** [comment]
It's strange that neither issue got a reply from the devs. This seems rather serious.

**Comment by ndas108** [comment]
I spent come time investigating. The bug seems to affect only the kopia-ui, not the main cli utility. The same bug also reveals itself in the process of mounting a snapshot in UI, which creates a new temporary folder in /tmp/ directory. That auto created folder has the wrong permission / timestamp. When using the cli utility no such problems occur, because the user has to manually create a directory before executing 'kopia mount' command. The same goes with 'kopia mount' command: it can not be executed without a preexisting directory.

**Comment by tessus** [comment]
What is your cli command to restore dir x from snapshot y to directory z?

And does z have the same perm/date as x? Because I don't seem to get it to work.

**Comment by ndas108** [comment]
Before you do 
```bash
kopia restore k95f1dc3e2db9651578dcbf0a20c65974/dwg/ /home/user/test/  # with optional flags
```
you should first 
```bash
kopia snapshot list
```
in order to list the needed snapshot ID (time-wise) and then
list the subfolders / files
```bash
kopia list k95f1dc3e2db9651578dcbf0a20c65974/ # optional -r flag for recursion
```
so that you know the needed absolute paths.  

That is documented in the https://kopia.io/docs/getting-started/ section. I had no permissions / timestamp issues when working with CLI (manual directory creation vs apperently buggy scripts in kopia-ui).

**Comment by tessus** [comment]
Ok, yep, those were the commands I used as well (although I usually don't use the last slash for source and dest). However, the timestamp is still not correct. The permissions are.

I was asking, because I thought I might have missed something, but you were using the same commands as I. I only ever compared the kopiaUI timestamp with the restored timestamp. Until now.

kopiaUI shows me the following timestamp for the dir: `8/30/2020, 10:35:41 PM`
the cli (kopia list) shows the following timestamp: `2020-10-18 21:51:29 EDT`

After the restore, the dir shows the timestamp: `2020-10-18 20:51`

This seems to be correct when comparing to the cli output, since I am currently in EST and not EDT. But the date in kopiaUI makes no sense, even when converting to UTC.
So the question is, what timestamp does the kopiaUI show?

**Comment by tessus** [comment]
Ok, I did some more debugging. I checked the source dir on my filesystem and it actually shows `2020-08-30 21:35`. This is the modified timestamp. So kopiaUI shows the modified timestamp.

Then I checked:
accessed: `2024-02-06 13:10`
created: `2019-01-05 23:25`

So now I am wondering what timestamp the cli shows and restores. Because none of them have anything to do with `2020-10-18 20:51`. 😕

**Comment by ndas108** [comment]
If I stick to defaults in both GUI (`Restore File Modification Time` checked) and CLI (`skip-times` flag is not used ) when performing restore my timestamps for subfolders / files are perfectly fine as they should be. The parent directory created by GUI is still corrupt though.  The same parent directory created by CLI (when not created manually in advance) is fine.

**Comment by ndas108** [comment]
I assume you are using one of the standard linux file systems in your repository, and it's not in the "cloud" via webdav, rclone, etc?

**Comment by tessus** [comment]
I think you are missing the point when it comes to my previous comment.

kopiaUI shows a different timestamp than `kopia list -l k95f1dc3e2db9651578dcbf0a20c65974 |grep testfolder`

While the kopiaUI and the source directory on my SSD show the modified timestamp for that directory, the cli does neither show the modified, accessed, or created timestamp. It shows a timestamp that has nothing to do with the directory. I checked with `stat` and other tools. So, the restored timestamp is the same as in the cli snapshot, but it's just the wrong timestamp.

> I assume you are using one of the standard linux file systems in your repository, and it's not in the "cloud" via webdav, rclone, etc?

Yes, but that shouldn't matter at all. The permissions/timestamps are stored in the repo (as metadata) not the filesystem. The repo consists of blobs and not 1:1 files.

**Comment by ndas108** [comment]
> I think you are missing the point when it comes to my previous comment.
> 
> kopiaUI shows a different timestamp than `kopia list -l k95f1dc3e2db9651578dcbf0a20c65974 |grep testfolder`
> 
> While the kopiaUI and the source directory on my SSD show the modified timestamp for that directory, the cli does neither show the modified, accessed, or created timestamp. It shows a timestamp that has nothing to do with the directory. I checked with `stat` and other tools. So, the restored timestamp is the same as in the cli snapshot, but it's just the wrong timestamp.
> 
Yes, I see now. I didn't experience that. Metadata in my restored subfolders / files is consistent with the original metadata and the repository metadata.  

> > I assume you are using one of the standard linux file systems in your repository, and it's not in the "cloud" via webdav, rclone, etc?
> 
> Yes, but that shouldn't matter at all. The permissions/timestamps are stored in the repo (as metadata) not the filesystem. The repo consists of blobs and not 1:1 files. 

Yeah, that's exactly the countermeasure against non-standard hosts as well as other data integrity problems. But I thought if its implementation was buggy somehow, "cloud" backend quirks could have theoretically interfered with the whole picture, and the info might have be useful to devs.

**Comment by ndas108** [comment]
I think that's why they still consider most "cloud" options (except for a number of tested ones) unstable.

**Comment by tessus** [comment]
I am just using a standard filesystem so no reason to follow that direction. It will only distract from the current issue.

**Comment by ndas108** [comment]
That's my point too. Less chance the devs might invoke the "bad cloud argument" so they might concentrate on the issue at hand :). Because I originally encountered the bug with my cloud repository, but made sure to reproduce it on my local one just in case.

**Comment by tessus** [comment]
This problem is not fixed with KopiaUI 0.22.0. The parent (created) directory still has the wrong permissions and date/time.

If the restore is done via the command line, the correct permissions are set.

**Comment by eliyen** [comment]
I'm experiencing a similar issue, possibly caused by the same bug, where the modified time of a parent directory is different between the GUI and CLI. I downloaded a file whose modified time was erroneously set to a future time. Directory structure is roughly as follows:

`/dir_1/dir_2/dir_3/future_time_file
`

When I take a snapshot of `/dir_1` via the GUI Snapshot Now button, I can see that `dir_1`, `dir_2`, and `dir_3` all show the correct modified times when browsing the snapshot directories in the GUI.

However, when I restore the snapshot via `kopia restore` on the command line, the modified times of `dir_1`, `dir_2`, and `dir_3` are all set to the modified time of `future_time_file`. I initially thought it was an issue with `kopia restore`, but when I tried `kopia list -l` on the snapshot, `dir_1`, `dir_2`, and `dir_3 ` all show up with the same future modified time. It seems the GUI and CLI are inconsistent in how they read file modified times.

In my case, the GUI is correct, and the CLI is wrong, which seems to be the opposite of what @tessus reported.
